FLEDGE_Key_Value_Server_trust_model.md

```diff
@@ -157,6 +157,40 @@ The initial implementation strategy for the key/value service is as follows:
 *   The TEE based key/value service implementation would be deployed on cloud service(s) which support needed security features. We envision the key/value service being capable of deployment with multiple cloud providers.
 
 
+# Support for user-defined functions (UDFs)
+
+Given ecosystem feedback, we plan to support user-defined functions (UDFs), to be loaded and executed inside the Key/Value Server.  This will allow more flexibility in how this server is used  and provide a server-side FLEDGE platform for code execution.  The loading mechanism has yet to be determined, and will be included in a future update to this explainer.
+
+By default we will provide a reference UDF that will do a lookup for the key and simply return the value.  Replacing this with a custom UDF is not needed if only basic lookup functionality is required.  (We expect the performance impact of the reference UDF to be negligible.)  The data store that holds the adtech-loaded data will be exposed to the UDF via new read-only APIs.
+
+This diagram shows how the interaction will work:
+
+![alt_text](assets/fledge_kv_server_custom_logic.png "UDF server-integration diagram")
+
+
+_The diagram shows the request handler inside the Key/Value Server being able to invoke a UDF which can use a new API to read data from the data store._
+
+
+## Design principles
+
+The following principles are necessary in order to preserve the trust model:
+
+
+
+*   _Sandbox_ - the custom code will be executed inside a sandbox that limits what it is allowed to do.  We’re currently looking at the [Open Source V8 engine](https://v8.dev/) inside [Sandbox2](https://developers.google.com/code-sandboxing/sandbox2), which has support for both JavaScript and Web Assembly (WASM).  Other suggestions are welcome!
+*   _No network, disk access, timers, or logging_ - this will be enforced using the sandbox, above.  This preserves the Key/Value Server’s principle of no side-effects and avoids leaking user data.  Coarse timers may be allowed but fine-grained timers are disallowed to help prevent covert channels (e.g. SPECTRE).
+*   _Individual request handling_ - Per the [FLEDGE explainer](https://github.com/WICG/turtledove/blob/main/FLEDGE.md#31-fetching-real-time-data-from-a-trusted-server), a request to the K/V Server may be for multiple keys.  The UDF will be called separately for each individual key, rather than one for all of the keys.  This ensures that each key is processed independently and prevents a group of keys from being used as a fingerprint for a user.
```

Comment:
This says "as a fingerprint for a user", but really it's preventing keys from being used as a _cross-site profile_ for a user.  (There's no restriction on assigned keys, so fingerprinting really isn't relevant here.)

Reply:
Updated